### PR TITLE
test: integration test for cron jobs service

### PR DESCRIPTION
**Pull request by matt2e**
Tests the real service with the real dal / postgres.
Executes jobs with a 2 second schedule 3 times.

### CODE REVIEW

File: backend/controller/cronjobs/cronjobs.go [hunk 1]
```diff
@@ -168,8 +168,8 @@ func (s *Service) CreatedOrReplacedDeloyment(ctx context.Context, newDeploymentK
 	_ = s.syncJobsWithNewDeploymentKey(ctx, optional.Some(newDeploymentKey))
 }
 
-// syncJobs is run periodically via a scheduled task
-func (s *Service) syncJobs(ctx context.Context) (time.Duration, error) {
+// SyncJobs is run periodically via a scheduled task
+func (s *Service) SyncJobs(ctx context.Context) (time.Duration, error) {
```
Comment (alecthomas) [inline backend/controller/cronjobs/cronjobs.go:172]: I think further to what we discussed earlier re. data model abstractions, we should also pull out all of the "logic" of the controller into a separate package such that the controller itself contains very little logic.

Not for this PR, just a note.

File: integration/cronjobs_integration_test.go [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+}
+
+type controller struct {
+	key      model.ControllerKey
+	dal      *db.DAL
+	clock    clock.Clock
+	cronJobs *cronjobs.Service
+}
+
+func TestServiceWithDB(t *testing.T) {
```
Comment (alecthomas) [inline integration/cronjobs_integration_test.go:48]: Per my comment in the other PR, can we factor out the common code in the two tests such that they're using the same logic?

**Review by alecthomas — COMMENTED** [review]
Per my comment on the other PR, the `integration.yml` workflow still needs to be updated to run this - currently it will not run at all.

**Comment by matt2e** [comment]
> Per my comment on the other PR, the `integration.yml` workflow still needs to be updated to run this - currently it will not run at all.

the yaml:
> echo "matrix={\"test\":$(jq -c -n '$ARGS.positional' --args $(grep '^func Test' integration/*_test.go | awk '{print $2}' | cut -d'(' -f1))}" >> "$GITHUB_OUTPUT"

I think this is already handled because I put this new file into the same directory so `integration/*_test.go` would be fine as is, same for the Justfile which also looks in the directory.

Not sure if this was better, or this should be moved back to the cronjobs package (which would also mean the `SyncJobs` func could be private again)

**Comment by alecthomas** [comment]
Ah sorry, I didn't see the directory change. But yeah, I think it would be preferable to share the code.

**Comment by matt2e** [comment]
new approach:
- `cronjobs_utils_test.go`: shared test code
- `cronjobs_test.go`: unit tests (will not be included when build tag `integration` is used)
- `cronjobs_integration_test.go`: integration tests (will only be included when build tag `integration` is used)
- Both `Justfile` and `integration.yml` add cronjobs directory to look for tests (with `integration` build tag)

**Comment by matt2e** [comment]
Is there a better approach to this?

**Comment by alecthomas** [comment]
> Is there a better approach to this?

Usually it's something like:

```go
func TestIntegration(t *testing.T) {
  d := dal.New()
  runTest(t, d)
}

func TestMocked(t *testing.T) {
  d := fakedal.New()
  runTest(t, d)
}

type DALSubset interface {
  // DAL methods we need.
}

func runTest(t *testing.T, d DALSubset) {
}
```

**Comment by matt2e** [comment]
I don't think that'd quite work here. The tests are slightly different because the real dal uses db's clock to make sure jobs aren't run too early, so the integration test has to be slow with longer `time.Sleep()` calls to match real time progression, which isn't really appropriate for the non integration test.
So i think the approach with different tests which focus on their differences and just call into common functionality is better here.

**Comment by matt2e** [comment]
Changed it so now we have a func that both the unit test and the integration test call into, just passing in different dals (mock vs real). The shared function also has to do some smarts around whether to progress in real time for the integration test (the db's clock ensures safety, which slows down the test) and the unit test (which can mock the clock to speed things up)

**Comment by matt2e** [comment]
(But there's still the separation of files and the build flag trick)

**Review by alecthomas — APPROVED** [review]

**Pull request merged**